File: src/swift/3_StructuralPatterns/4_Decorator/StaticDecorator.swift
import Foundation

protocol Shape : CustomStringConvertible
{
  init() // requited for construction
  var description: String { get }
}

class Circle : Shape
{
  private var radius: Float = 0

  required init() {}
  init(_ radius: Float)
  {
    self.radius = radius
  }

  func resize(_ factor: Float)
  {
    radius *= factor
  }

  public var description: String
  {
    return "A circle of radius \(radius)"
  }
}

class Square : Shape
{
  private var side: Float = 0

  required init() {}
  init(_ side: Float)
  {
    self.side = side
  }

  public var description: String
  {
    return "A square with side \(side)"
  }
}

class ColoredShape<T> : Shape where T : Shape
{
  private var color: String = "black"
  private var shape: T = T()

  required init() {}
  init(_ color: String)
  {
    self.color = color
  }

  public var description: String
  {
    return "\(shape.description) has the color \(color)"
  }
}

class TransparentShape<T> : Shape where T : Shape
{
  private var transparency: Float = 0
  private var shape: T = T()
  
  required init(){}
  init(_ transparency: Float)
  {
    self.transparency = transparency
  }

  public var description: String
  {
    return "\(shape.description) has transparency \(transparency*100)%"
  }
}

func main()
{
  let blueCircle: ColoredShape<Circle> = ColoredShape<Circle>("blue")
  print(blueCircle)

  // only transparency propagates, color does not
  let blackHalfSquare = TransparentShape<ColoredShape<Square>>(0.4)
  print(blackHalfSquare)
}

main()

File: src/java/3_StructuralPatterns/4_Decorator/StaticDecorator.java
package com.activemesa.structural.decorator.statik;

import java.util.function.Supplier;

interface Shape
{
    String info();
}

class Circle implements Shape
{
    private float radius;

    Circle(){}

    public Circle(float radius)
    {
        this.radius = radius;
    }

    void resize(float factor)
    {
        radius *= factor;
    }

    @Override
    public String info()
    {
        return "A circle of radius " + radius;
    }
}

class Square implements Shape
{
    private float side;

    public Square()
    {
    }

    public Square(float side)
    {
        this.side = side;
    }

    @Override
    public String info()
    {
        return "A square with side " + side;
    }
}

// we are NOT altering the base class of these objects
// cannot make ColoredSquare, ColoredCircle

class ColoredShape<T extends Shape> implements Shape {
    private Shape shape;
    private String color;

    public ColoredShape(Supplier<? extends T> ctor, String color) {
        shape = ctor.get();
        this.color = color;
    }

    @Override
    public String info() {
        return shape + " has the color " + color;
    }
}

class TransparentShape<T extends Shape> implements Shape
{
    private Shape shape;
    private int transparency;

    public TransparentShape(Supplier<? extends T> ctor, int transparency)
    {
        shape = ctor.get();
        this.transparency = transparency;
    }

    @Override
    public String info() {
        return shape + " has " + transparency + "% transparency";
    }
}

class StaticDecoratorDemo {
    public static void main(String[] args) {
        Circle circle = new Circle(10);
        System.out.println(circle.info());

        ColoredShape<Square> blueSquare = new ColoredShape<>(() -> new Square(20), "blue");
        System.out.println(blueSquare.info());

        // ugly!
        TransparentShape<ColoredShape<Circle>> myCircle =
          new TransparentShape<>(() ->
              new ColoredShape<>(() -> new Circle(5), "green"), 50
          );
        System.out.println(myCircle.info());
        // cannot call myCircle.resize()
    }
}
